### ISSUE TITLE
[java] CloseResource: False negative for resource leaks in anonymous objects and loops

### ISSUE THREAD

**Issue by leemeii** [pmd/pmd#6494]
**Affects PMD Version:** 7.20.0

**Rule:**[CloseResource](https://pmd.github.io/pmd/pmd_rules_java_errorprone.html#closeresource)

**Description:**
The rule CloseResource fails to detect resource leaks when the resource is created as a temporary (anonymous) object in method chains or repeatedly in loop conditions, even though these scenarios lead to unclosed resources like Scanner.

**Code Sample demonstrating the issue:**

```java
import java.util.Scanner;  
import java.io.ByteArrayInputStream;  

public class ResourceLeakDemo {  

    // Resource source: Creates a new Scanner each time it's called  
    public Scanner getNewScanner() {  
        // Actually creates a resource that needs to be closed  
        return new Scanner(new ByteArrayInputStream("hello".getBytes()));  
    }  

    // Case A: Leak that the tool can detect  
    public void detectedLeak() {  
        Scanner s = getNewScanner();   //<- reported correctly(TP)
        System.out.println(s.next());  
         
    }  

    // Case B: Serious leak that the tool misses (false negative)  
    public void silentLeak() {  
        // Direct method chain call.  
        // This creates a Scanner object but doesn't assign it to any variable.  
        System.out.println(getNewScanner().next());  //<-should report (FN)
          
    }  

    // Case C: Leak in loop variant code (most severe)  
    public void loopLeak() {  
        // If this loop runs 1000 times, it will leak 1000 Scanners  
        while (getNewScanner().hasNext()) {   //<-should report (FN)
            // Each hasNext() check creates a new Scanner, which can never be closed  
            break;   
        }  
    }  
}  
```

**Expected outcome:**

PMD should report a CloseResource violation for the Scanner creations in silentLeak (line with getNewScanner().next()) and loopLeak (line with while condition), as these create unclosed resources, similar to the detected case in detectedLeak. This is a false-negative.

**Running PMD through:** *CLI*

**Comment by blacelle** [comment]
I add a related case to this one (not to clutter this report, but to provide some related behavior which looks awkward).

I consider a class keeping a reference to a typically long-lived `AutoCloseable`: an `ExecutorService`.

The following reports for `CloseResource`:
```
ExecutorService executor = wrapper.getExecutorService();
executor.submit(() -> {...});
```

while the following has no report:
```
wrapper.getExecutorService().submit(() -> {...});
```

In my case, it feels more like a false-positive, as `ExecutorService executor = wrapper.getExecutorService()` is not actually opening a resource, but I suppose PMD can not know.

Still, the inconsistency if the `Closeable` is allocated in a variable looks fishy (which is very similar to the original post).

I'm not sure how it relates to the `false-negative` case, but we may benefit from a way to clarify when a method returning a `Closeable` is opening a resource or not (e.g. through an annotation on the method, instead of annotating each and every call to it).
